tap_google_ads/streams.py

```diff
@@ -330,22 +380,60 @@ def sync(self, sdk_client, customer, stream, config, state): # pylint: disable=u
+        last_pk_fetched = singer.get_bookmark(state,
+                                              stream["tap_stream_id"],
+                                              customer["customerId"]) or {}
+
+        query = create_core_stream_query(resource_name, selected_fields, last_pk_fetched.get('last_pk_fetched'), self.filter_param)
         try:
             response = make_request(gas, query, customer["customerId"])
         except GoogleAdsException as err:
             LOGGER.warning("Failed query: %s", query)
             raise err
 
-        with Transformer() as transformer:
-            # Pages are fetched automatically while iterating through the response
-            for message in response:
-                json_message = google_message_to_json(message)
-                transformed_message = self.transform_keys(json_message)
-                record = transformer.transform(transformed_message, stream["schema"], singer.metadata.to_map(stream_mdata))
+        with metrics.record_counter(stream_name) as counter:
+            with Transformer() as transformer:
+                # Pages are fetched automatically while iterating through the response
+                for message in response:
+                    json_message = google_message_to_json(message)
+                    transformed_message = self.transform_keys(json_message)
+
+                    # Skip already synced record in interrupted sync.
+                    if not should_sync_record(transformed_message, self.filter_param, self.primary_keys, last_pk_fetched.get('last_pk_fetched')):
+                        continue
+
+                    record = transformer.transform(transformed_message, stream["schema"], singer.metadata.to_map(stream_mdata))
+
+                    singer.write_record(stream_name, record)
+                    counter.increment()
+
+                    # Write state(last_pk_fetched) using primary key(id) value for core streams after DEFAULT_PAGE_SIZE records
+                    if counter.value % DEFAULT_PAGE_SIZE == 0 and self.filter_param:
+                        bookmark_value = {}
+                        for param_index, param_value in enumerate(self.primary_keys):
+                            bookmark_value[self.filter_param[param_index]] = record[param_value]
 
-                singer.write_record(stream_name, record)
+                        singer.write_bookmark(state, stream["tap_stream_id"], customer["customerId"], {'last_pk_fetched': bookmark_value})
+                        singer.write_state(state)
 
+                        LOGGER.info("Write state for stream: %s, value: %s", stream_name, bookmark_value)
+
+        # Flush the state for core streams if sync is completed
+        if stream["tap_stream_id"] in state.get('bookmarks', {}):
+            state['bookmarks'].pop(stream["tap_stream_id"])
+            singer.write_state(state)
+
+def should_sync_record(record, filter_param, pks, last_pk_fetched):
```

Comment:
Add unit tests for above function.

Reply:
Added unittest cases for `should_sync_record` method.